emote/models/ensemble.py

```diff
@@ -0,0 +1,351 @@
+        )
+        self.min_logvar = nn.Parameter(
+            -10 * torch.ones(1, out_size), requires_grad=learn_logvar_bounds
+        )
+        self.max_logvar = nn.Parameter(
+            0.5 * torch.ones(1, out_size), requires_grad=learn_logvar_bounds
+        )
+        self.logvar_loss_weight = 0.01
+        self.apply(truncated_normal_init)
+        self.to(self.device)
+
+    def _default_forward(
+        self, x: torch.Tensor, only_elite: bool = False, **_kwargs
+    ) -> tuple[torch.Tensor, Optional[torch.Tensor]]:
+        x = self.hidden_layers(x)
+        mean_and_logvar = self.mean_and_logvar(x)
+        mean = mean_and_logvar[..., : self.out_size]
+        logvar = mean_and_logvar[..., self.out_size :]
+        logvar = self.max_logvar - F.softplus(self.max_logvar - logvar)
+        logvar = self.min_logvar + F.softplus(logvar - self.min_logvar)
+        return mean, logvar
+
+    def _forward_from_indices(
+        self, x: torch.Tensor, model_shuffle_indices: torch.Tensor
+    ) -> tuple[torch.Tensor, Optional[torch.Tensor]]:
+        _, batch_size, _ = x.shape
+
+        num_models = (
+            len(self.elite_models) if self.elite_models is not None else len(self)
+        )
+        shuffled_x = x[:, model_shuffle_indices, ...].view(
+            num_models, batch_size // num_models, -1
+        )
+
+        mean, logvar = self._default_forward(shuffled_x, only_elite=True)
+        # note that mean and logvar are shuffled
+        mean = mean.view(batch_size, -1)
+        mean[model_shuffle_indices] = mean.clone()  # invert the shuffle
+
+        if logvar is not None:
+            logvar = logvar.view(batch_size, -1)
+            logvar[model_shuffle_indices] = logvar.clone()  # invert the shuffle
+
+        return mean, logvar
+
+    def _forward_ensemble(
+        self, x: torch.Tensor
+    ) -> tuple[torch.Tensor, Optional[torch.Tensor]]:
+        """
+        Valid propagation options are:
+            - "random_model": for each output in the batch a model will be chosen at random.
+              This corresponds to TS1 propagation in the PETS paper.
+            - "fixed_model": for output j-th in the batch, the model will be chosen according to
+              the model index in `propagation_indices[j]`. This can be used to implement TSinf
+              propagation, described in the PETS paper.
+            - "expectation": the output for each element in the batch will be the mean across
+              models.
+        """
+        if self.propagation_method is None:
+            mean, logvar = self._default_forward(x, only_elite=False)
+            if self.num_members == 1:
+                mean = mean[0]
+                logvar = logvar[0] if logvar is not None else None
+            return mean, logvar
+        assert x.ndim == 2
+        model_len = (
+            len(self.elite_models) if self.elite_models is not None else len(self)
+        )
+        if x.shape[0] % model_len != 0:
```

Comment:
Maybe should be assert? :-)

Reply:
Fixed! Actually it was not needed at all.